File: DateTestObjC/DateTestObjC/Date Manager/DateManager.m
//
//  DateManager.m
//  DateTestObjC
//
//  Created by Andy Shephard on 15/05/2017.
//  Copyright © 2017 Andy Shephard. All rights reserved.
//

#import "DateManager.h"

static NSInteger kOneDayInSeconds = 86400;
static NSString *kISO8601DateFormat = @"yyyy-MM-dd'T'HH:mm:ssZZZZZ";
static NSString *kDayFormat = @"EEEE";

NSString *kKeyISO8601Date = @"ISO8601Date";
NSString *kKeyDayName = @"DayName";
NSString *kKeyDifference = @"Difference";

@interface DateManager()
@property (nonatomic, strong) NSDateFormatter *dateFormatter;
@property (nonatomic, strong) NSDateFormatter *dayFormatter;
@end

@implementation DateManager

+ (instancetype)sharedManager {
	static DateManager *sharedMyManager = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		sharedMyManager = [[self alloc] init];
	});
	return sharedMyManager;
}

- (id)init {
	if (self = [super init]) {
		// Set |dateFormatter| to be ISO8601 compliant.
		_dateFormatter = [[NSDateFormatter alloc] init];
		NSLocale *enUSPosixLocale = [[NSLocale alloc] initWithLocaleIdentifier:@"en_US_POSIX"];
		_dateFormatter.locale = enUSPosixLocale;
		_dateFormatter.dateFormat = kISO8601DateFormat;
		
		// Set |dayFormatter| to return the Day Name.
		_dayFormatter = [[NSDateFormatter alloc] init];
		_dayFormatter.dateFormat = kDayFormat;
		
		// Set |today| as today's date (right now).
		_now = [NSDate date];
		_today = [_dateFormatter stringFromDate:_now];
	}
	return self;
}

- (void)calculateValuesFromSelectedDate:(NSDate *)date
							 completion:(void (^)(NSDictionary *results)) completionBlock {
	
	// 1. Selected date in NSString.
	NSString *iso8601String = [_dateFormatter stringFromDate:date];
	
	// 2. Day name in NSString.
	NSString *dayNameString = [_dayFormatter stringFromDate:date];
	
	// 3. Difference from selected date to today's date in NSString.
	NSString *difference = [self daysBetweenSelectedDateAndNow:date];
	
	// Add all values into a NSDictionary object and return them via the
	// completion handler.
	NSDictionary *resultsDict = [[NSDictionary alloc] initWithObjectsAndKeys:iso8601String, kKeyISO8601Date, dayNameString, kKeyDayName, difference, kKeyDifference, nil];
	
	completionBlock(resultsDict);
}

- (NSString *)daysBetweenSelectedDateAndNow:(NSDate *)date {
	
	// Get timestamp difference between |now| and
	// the selected date.
	NSTimeInterval selectedDateInSeconds = [date timeIntervalSinceDate:_now];
	
	// Calculate difference in days (integer).
	double timeIntervalInteger = selectedDateInSeconds;
	NSInteger difference = [self roundToDays:timeIntervalInteger] / kOneDayInSeconds;
	
	// Return the NSInteger value as an NSString.
	return [NSString stringWithFormat:@"%li", (long)difference];
}

- (NSInteger)roundToDays:(double)x {
	
	return kOneDayInSeconds * floor((x / kOneDayInSeconds) + 0.5);
	
//	return kOneDayInSeconds * floor((x / kOneDayInSeconds));
}

@end

File: DateTestSwift/DateTestSwift/Date Manager/DateManager.swift
//
//  DateManager.swift
//  DateTestSwift
//
//  Created by Andy Shephard on 15/05/2017.
//  Copyright © 2017 Andy Shephard. All rights reserved.
//

import Foundation

let kOneDayInSeconds = 86400

let kISO8601DateFormat = "yyyy-MM-dd'T'HH:mm:ssZZZZZ"
let kDayFormat = "EEEE"

let kKeyISO8601Date = "ISO8601Date"
let kKeyDayName = "DayName"
let kKeyDifference = "Difference"

public class DateManager {
	
	let dateFormatter: DateFormatter
	let dayFormatter: DateFormatter
	
	let now: Date
	let today: String
	
	static let sharedInstance = DateManager()
	
	private init() {
		// Set |dateFormatter| to be ISO8601 compliant.
		dateFormatter = DateFormatter.init()
		let enUSPosixLocale = NSLocale(localeIdentifier: "en_US_POSIX")
		dateFormatter.locale = enUSPosixLocale as Locale!
		dateFormatter.dateFormat = kISO8601DateFormat
		
		// Set |dayFormatter| to return the Day Name.
		dayFormatter = DateFormatter.init()
		dayFormatter.dateFormat = kDayFormat
		
		// Set |today| as today's date (right now).
		now = Date.init()
		today = dateFormatter.string(from: now)
	}
	
	func calculateValuesFromSelectedDate(_ date: Date, completion: (_ result: [String:String]) -> Void) {
	
		// 1. Selected date in String.
		let iso8601String: String = dateFormatter.string(from: date)
		
		// 2. Day name in String.
		let dayNameString: String = dayFormatter.string(from: date)
		
		// 3. Difference from selected date to today's date in String.
		let difference: String = daysBetweenSelectedDateAndNow(date)
		
		// Add all values into a Dictionary object and return them via the
		// completion handler.
		let resultsDict:[String:String] = [kKeyISO8601Date:iso8601String, kKeyDayName:dayNameString, kKeyDifference:difference]
		
		completion(resultsDict)
	}
	
	func daysBetweenSelectedDateAndNow(_ date: Date) -> String {
		
		// Get timestamp difference between |now| and
		// the selected date.
		let selectedDateInSeconds: TimeInterval = date.timeIntervalSince(now)
		
		// Calculate difference in days (integer).
		let timeIntervalInteger: Double = Double(selectedDateInSeconds)
		let difference: Int = roundToDays(x: timeIntervalInteger) / kOneDayInSeconds
		
		// Return the Int value as a String.
		return String(difference)
	}
	
	func roundToDays(x : Double) -> Int {
		return kOneDayInSeconds * Int(round(x / Double(kOneDayInSeconds)))
	}

}

extension Double {
	func roundToDays() -> Int{
		return 86400 * Int(Darwin.round(self / 86400.0))
	}
}
